[TOPIC-GPIO] gpio: Update imx gpio driver to use new gpio api

stanislav-poboril opened the pull request:
Updates the NXP i.MX gpio driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle and pin_interrupt_configure functions recently
added to the gpio api.

Assumes the gpio api layer handles translating logical flags to physical
flags.

Tested with:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_api_1pin
- tests/drivers/gpio/gpio_basic_api

On boards:
- udoo_neo_full_m4

The udoo_neo_full_m4 board does not have a user button, so I temporarily modified the samples/basic/button code and board dts file to use second pin to simulate the button.

The test tests/drivers/gpio/gpio_api_1pin skips most of the tests, since it relies on pin to be configured as both input/output at the same time. i.MX gpio can be either input or output. However it is possible to read back the data from the output pins, so I temporarily modified the code to be able to test some of the cases.

I don't have colibri_imx7d_m4 board, only verified that the samples/basic/blinky and samples/basic/button examples build. The test tests/drivers/gpio/gpio_api_1pin compiles but does not link since it does not fit into board's flash.

stanislav-poboril committed:
dts: nxp_imx6sx_m4: fix gpio5 int num

Fix interrupt number for gpio5

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
gpio: Update imx gpio driver to use new gpio api

Updates the imx gpio driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle and pin_interrupt_configure functions recently
added to the gpio api.

Assumes the gpio api layer handles translating logical flags to physical
flags.

Tested with:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_api_1pin
- tests/drivers/gpio/gpio_basic_api

On boards:
- udoo_neo_full_m4

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

galak commented:
Are you able to test with the `tests/drivers/gpio/gpio_basic_api`.  It will require a wire to loop back a pair of pins.  This test covers more so feel pretty confident if we're able to pass it.

galak reviewed (dismissed):
Should convert boards/arm/96b_meerkat96/96b_meerkat96.dts & boards/arm/warp7_m4/warp7_m4.dts

stanislav-poboril committed:
tests/drivers/gpio: Added udoo_neo_full_m4 board to gpio_basic_api test

Added configuration for the udoo_neo_full_m4 board and
the initialization of its GPIO test pins via IOMUX controller.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril commented:
> Are you able to test with the `tests/drivers/gpio/gpio_basic_api`. It will require a wire to loop back a pair of pins. This test covers more so feel pretty confident if we're able to pass it.

Updated the code, added the udoo_neo_full_m4 board configuration and pin mux code into the `tests/drivers/gpio/gpio_basic_api` sample. It is passing now.

> Should convert boards/arm/96b_meerkat96/96b_meerkat96.dts & boards/arm/warp7_m4/warp7_m4.dts

Done.

mnkp commented on drivers/gpio/gpio_imx.c:
```diff
@@ -106,46 +117,162 @@ static int imx_gpio_read(struct device *dev,
+		return -ENOTSUP;
+	}
+
+	if ((mode == GPIO_INT_MODE_LEVEL) && (trig == GPIO_INT_TRIG_BOTH)) {
```
I'll add this check as an assert to GPIO API. Let's remove it from the driver.

carlescufi commented:
@stanislav-poboril 
The topic branch has been rebased. Could you please rebase this PR against it?

stanislav-poboril commented:
> 
> 
> @stanislav-poboril
> The topic branch has been rebased. Could you please rebase this PR against it?

Rebased and retested.
